### ISSUE TITLE
the loss of train is strange

### ISSUE THREAD

**Issue by Soda-Wong** [AlexeyAB/darknet#276]
I meet a strange problem. 
I used my dataset to train on yolo-voc.2.0 as you said. 
but the loss became small quickly and the obj is low all the time.
I try to change different learn-rate, but no effects.

there is just one class for ship and 2500 images in my dataset.
can someone help me?
<img width="504" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/33228137-a4add24c-d179-11e7-92ff-ff435bbdc9d0.png">
<img width="918" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/33228139-adbaf63a-d179-11e7-9a2b-b18c6a597770.png">

**Comment by shizukachan** [comment]
looks good to me, test it

**Comment by AlexeyAB** [comment]
Just try to train about 2000 iterations, and validate/recall it. How much IoU can you get by using `darknet.exe detector recall ...`? https://github.com/AlexeyAB/darknet#when-should-i-stop-training

**Comment by Soda-Wong** [comment]
@shizukachan @AlexeyAB  I use the darknet.exe detector recall ... and what I get is 0.
<img width="648" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/33236594-e94e08aa-d21d-11e7-83b7-36cf97c8244b.png">
and I can't any ship in my image even the thresh is 0.01.
there is other things I can do?

**Comment by AlexeyAB** [comment]
Show you .data-file, and show several lines from your train.txt and valid.txt.
Also can you show full command lines which did you use for train and recall?

**Comment by Soda-Wong** [comment]
@AlexeyAB ok!
here is my data-flie:
<img width="258" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/33239942-d046a632-d272-11e7-822d-571768c2d148.png">
and one line of the train.txt: 
`D:/darknet-master/build/darknet/x64/data/voc/VOCdevkit/VOC2007/JPEGImages/000001.jpg`
and one line of the valid.txt: 
 `D:/darknet-master/build/darknet/x64/data/voc/VOCdevkit/VOC2007/JPEGImages/000006.jpg`

and there is my command lines for train:
`>darknet.exe detector train data/voc.data yolo-voc.2.0.cfg darknet19_448.conv.23`
for recall:
`darknet.exe detector recall data/voc.data yolo-voc.2.0.cfg backup/ship/yolo-voc_final.weights`

**Comment by AlexeyAB** [comment]
@Soda-Wong Everything looks correct.
* Can you show your `yolo-voc.2.0.cfg`?
* How many iterations did you train?

**Comment by Soda-Wong** [comment]
@AlexeyAB here is my `yolo-voc.2.0.cfg`:
`[net]
batch=64
subdivisions=8
height=416
width=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.0001
max_batches = 16000
policy=steps
steps=100,4000,15000
scales=.1,1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky


#######

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[route]
layers=-9

[reorg]
stride=2

[route]
layers=-1,-3

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear

[region]
anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52
bias_match=1
classes=1
coords=4
num=5
softmax=1
jitter=.2
rescore=1

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .5
random=0
`
 As described in the file,I have trained for 16000 iterations.
Now I'm trying the yolo-voc.cfg,I will give you the results when the train is done.:)

**Comment by Soda-Wong** [comment]
@AlexeyAB the train is done but the result is the same as the yolo-voc.2.0.cfg.
the defferent in loss is that loss gradually dropped to 0.1 when the iteration is 16000.

**Comment by AlexeyAB** [comment]
Try to take weights for 2000 iterations: `darknet.exe detector recall data/voc.data yolo-voc.2.0.cfg backup/ship/yolo-voc_2000.weights`

**Comment by Soda-Wong** [comment]
It's still all 0.
Have you fixed your code since August？Because I download your code in August?

**Comment by AlexeyAB** [comment]
Yes, I fixed some bugs (that disrupt the learning).
But as I see your fork at 8 November: https://github.com/Soda-Wong/darknet/commits/master

Try to update your code, recompile and train again.

**Comment by Soda-Wong** [comment]
@AlexeyAB Sorry. I can't catch you exactly. You mean I update my code according to what you fixed after 8 November?But I find some different between the other fixed thing?

**Comment by AlexeyAB** [comment]
@Soda-Wong Just update your code on the computer from this my repo. Or try to use this tested release: https://github.com/AlexeyAB/darknet/releases

Then recompile code, and train from the begining.

**Comment by Soda-Wong** [comment]
@AlexeyAB thank you!!!
I update and  recompile my code. than the model I trained can detect the regions!
But here is some detection question. My ship sizes vary from small to large in images：
<img width="1131" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/33307503-551f853a-d3dc-11e7-9e6f-97f7011b4260.png">



now I use 4000 iterations model to detect them.The prediction of small ship is only 13% . The large one is 98% and some prediction boxed overlap.
So I think that there is some strategy to improve this situation?

And I will train for 16000 iterations first and see how is the results of 16000 iteration model.

**Comment by AlexeyAB** [comment]
@Soda-Wong What resolution of you images?

* You can set network resolution `width=832 height=832` in your cfg-file: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

* You can add more anchors, the same anchors + the same anchors scaled X 2: https://github.com/AlexeyAB/darknet/blob/617cf313ccb1fe005db3f7d88dec04a04bd97cc2/cfg/yolo-voc.2.0.cfg#L228-L232
```
anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52, 2,2,  6,8,  12,22,  18,10,  32,20
num=10
```

**Comment by Soda-Wong** [comment]
@AlexeyAB the resolution of my images is 1920*1080.
Do the values of anchors stand for that their resolution is the value multiply 13？
I am not sure.
And I have a idea that I can calculate the ten highest frequency anchors as initial anchors, will this work?

**Comment by AlexeyAB** [comment]
@Soda-Wong The most of images on VOC2012 has resolution 500x375 and 375x500. 
But anchors depends on network resolution instead of image resolution. So, because anchors in the yolo-voc.2.0.cfg calculated for 416x416 network resolution, then if you use the same 416x416 you can use the same anchors.

But the best way is to re-calculate your own acnhors for your network resolution and your own images.

**Comment by Soda-Wong** [comment]
@AlexeyAB sry for being so late to reply. I want to calculate my own anchor.The resolution of my images is 1920*1080 and my  network resolution is 416x416.
First I use K-means to get the first six widths and heights of my ships.
Then I use the formulas:width/(1920/416)/32 and height/(1920/416)/32 to calculate my anchor.
which I get is anchors = 1.09,0.65,  3.12,1.34,  5.51,1.92,  8.28,2.67,  11.91,3.11
then I use this to do the recall and get the bad results.So I realized I need to train first using these anchors.
And I get almost the same mAP with the original ahchors.
but the test results is strange.
This is the results with the original anchors:
<img width="486" alt="screenshot" src="https://user-images.githubusercontent.com/22865487/34410251-5f96be5a-eb94-11e7-9576-79c79bdd62dc.png">
This is the results with my own anchors:
<img width="489" alt="screenshot2" src="https://user-images.githubusercontent.com/22865487/34410262-731c3c0c-eb94-11e7-9d72-9c794a396107.png">
Can you tell my why?Thx very much!!

**Comment by AlexeyAB** [comment]
@Soda-Wong 

* Did you train with your own anchors, or did you train with default anchors and only test with your own anchors?
* You should train with your own anchors.

**Comment by Soda-Wong** [comment]
@AlexeyAB Yes. I train with my own anchors and the mAP is almost the same with the weights with default anchors.

**Comment by AlexeyAB** [comment]
@Soda-Wong 

* Did you calculate `width` of objects using k-means on images resized to 416x416 resolution?

----

> Then I use the formulas:width/(1920/416)/32 and height/(1920/416)/32 to calculate my anchor.

So, for example, if `width_obj = 100`, then:
* for 1920x1920: `anchor_w = width_obj/(1920/416)/32 = 100/((1920/416)*32) = 100/(4.61*32) = 0.677`
* for 416x416: `anchor_w = width_obj/(416/416)/32 = 100/32 = 3.125`

Did you do this?

----

I think if you calculated `width` of objects using k-means on images resized to 416x416, then for resolution 1920x1920 it should be `anchor_w = width_obj/(416/1920)/32`
So for example for `width_obj = 100` should be `anchor_w = 100/(416/1920)/32 = 100/6.9 = 14.5`
If I'm not wrong anywhere.

**Comment by Soda-Wong** [comment]
Sorry for wrong writing of formulas:
`width/(1920/416)/32 and height/(1080/416)/32. `
But it's no big deal.
the `width` is the value of original image which is 1920x1080, so I use this formula.

**Comment by Ahagpp** [comment]
@Soda-Wong 
`width/(1920/416)/32`  
why there divided by 32？

**Comment by Soda-Wong** [comment]
@Ahagpp 32 is the downsampling of the network
